### PR TITLE
common/buffer: close pipe fd if set nonblocking fails.

### PR DESCRIPTION
**Pull request by donglinpeng**
Signed-off-by: donglinpeng <donglinpeng@cmss.chinamobile.com>

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -540,6 +540,7 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
       if (r < 0) {
 	bdout << "raw_pipe: error setting nonblocking flag on temp pipe: "
 	      << cpp_strerror(r) << bendl;
+	close_pipe(tmpfd);
```
Comment (tchaikov) [inline src/common/buffer.cc]: might want to consolidate all the "close_pipe()` calls in the exit path to a single one with https://github.com/ceph/ceph/blob/master/src/include/scope_guard.h .

Comment (donglinpeng) [inline src/common/buffer.cc]: Nice. I will resubmit code after building test with scope_guard.h.

Comment (donglinpeng) [inline src/common/buffer.cc]: Hi, I try to consolidate all the "close_pipe()` calls in the exit path to a single one with scope_guard.h, but I encountered one test case failure. I am not sure how to use scope_guard.h. Could you give me some suggests, thank you.

**Comment by tchaikov** [comment]
retest this please.

File: src/common/buffer.cc [hunk 2]
```diff
@@ -536,6 +537,7 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
 	      << bendl;
 	throw error_code(r);
       }
+      auto sg = make_scope_guard([&] { close_pipe(tmpfd); });
```
Comment (tchaikov) [inline src/common/buffer.cc]: you change should work. but i'd suggest not capture `this` if not necessary. maybe we can mark `close_pipe()` a static method, like:
```diff
diff --git a/src/common/buffer.cc b/src/common/buffer.cc
index 7ca3a66d29..5113e8e5d8 100644
--- a/src/common/buffer.cc
+++ b/src/common/buffer.cc
@@ -515,7 +515,7 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
       return 0;
     }

-    void close_pipe(int *fds) {
+    static void close_pipe(const int *fds) {
       if (fds[0] >= 0)
        VOID_TEMP_FAILURE_RETRY(::close(fds[0]));
       if (fds[1] >= 0)
@@ -537,7 +537,7 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
              << bendl;
        throw error_code(r);
       }
-      auto sg = make_scope_guard([&] { close_pipe(tmpfd); });
+      auto sg = make_scope_guard([=] { close_pipe(tmpfd); });
       r = set_nonblocking(tmpfd);
       if (r < 0) {
        bdout << "raw_pipe: error setting nonblocking flag on temp pipe: "
```

Comment (donglinpeng) [inline src/common/buffer.cc]: Thank you very much. I have resubmitted codes according to you advice.

**Pull request merged**